h2o-algos/src/main/java/hex/gam/GAM.java

```diff
@@ -0,0 +1,251 @@
+package hex.gam;
+
+import hex.DataInfo;
+import hex.ModelBuilder;
+import hex.ModelCategory;
+import hex.gam.GAMModel.GAMParameters;
+import hex.gam.MatrixFrameUtils.GamUtils;
+import hex.glm.GLM;
+import hex.glm.GLMModel;
+import hex.glm.GLMModel.GLMParameters;
+import water.DKV;
+import water.Key;
+import water.MemoryManager;
+import water.Scope;
+import water.exceptions.H2OModelBuilderIllegalArgumentException;
+import water.fvec.Frame;
+import water.fvec.Vec;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static hex.gam.MatrixFrameUtils.GamUtils.*;
+
+
+public class GAM extends ModelBuilder<GAMModel, GAMModel.GAMParameters, GAMModel.GAMModelOutput> {
+
+  @Override
+  public ModelCategory[] can_build() {
+    return new ModelCategory[]{ModelCategory.Regression};
+  }
+
+  @Override
+  public boolean isSupervised() {
+    return true;
+  }
+
+  @Override
+  public BuilderVisibility builderVisibility() {
+    return BuilderVisibility.Experimental;
+  }
+
+  @Override
+  public boolean havePojo() {
+    return false;
+  }
+
+  @Override
+  public boolean haveMojo() {
+    return false;
+  }
+
+  public GAM(boolean startup_once) {
+    super(new GAMModel.GAMParameters(), startup_once);
+  }
+
+  public GAM(GAMModel.GAMParameters parms) {
+    super(parms);
+    init(false);
+  }
+
+  public GAM(GAMModel.GAMParameters parms, Key<GAMModel> key) {
+    super(parms, key);
+    init(false);
+  }
+  
+  @Override
+  public void init(boolean expensive) {
+    super.init(expensive);
+    if (expensive) {  // add custom check here
+      if (error_count() > 0)
```

Comment:
Should the error check be done after the errors below are also added ? I've checked e.g. GBM and we're doing it like that over there as well, but I admit I don't know the reason why.

Reply:
Pavel:

This is done to save a little time.  There are certain parameters that are universal to all models.  If you have error in those checks, we should probably stopped right away instead of plunging ahead.  As long as we check again before building model, it should be okay.  However, I know you would like us to use consistent implementation for all algos.  That is also very nice.

Reply:
I see, ok, thanks.